ext/mbstring/tests/eucjp_2004_encoding.phpt

```diff
@@ -0,0 +1,77 @@
+--TEST--
+Exhaustive test of EUC-JP-2004 encoding verification and conversion
+--EXTENSIONS--
+mbstring
+--SKIPIF--
+<?php
+if (getenv("SKIP_SLOW_TESTS")) die("skip slow test");
+?>
+--FILE--
+<?php
+srand(200); /* Make results consistent */
+include('encoding_tests.inc');
+mb_substitute_character(0x25); // '%'
+
+$validChars = array(); /* EUC-JP-2004 string -> UTF-32BE */
+$fromUnicode = array(); /* UTF-16BE -> EUC-JP-2004 */
+$fp = fopen(__DIR__ . '/data/EUC-JP-2004.txt', 'r+');
+while ($line = fgets($fp, 256)) {
+  if ($line[0] == '#')
+    continue;
+
+  $codepoint2 = null;
+  if (sscanf($line, "0x%x\tU+%x+%x", $bytes, $codepoint1, $codepoint2) >= 2) {
+    if ($bytes < 256)
+      $eucjp = chr($bytes);
+    else if ($bytes <= 0xFFFF)
+      $eucjp = pack('n', $bytes);
+    else
+      $eucjp = chr($bytes >> 16) . pack('n', $bytes & 0xFFFF);
+
+    if ($codepoint2) {
+      $validChars[$eucjp] = pack('NN', $codepoint1, $codepoint2);
+    } else {
+      $validChars[$eucjp] = pack('N', $codepoint1);
+      if ($codepoint1 <= 0xFFFF)
+        $fromUnicode[pack('n', $codepoint1)] = $eucjp;
+    }
+  }
+}
+
+/* Convert 0xA1B1 to U+FFE3 (FULLWIDTH MACRON), not U+203E (OVERLINE) */
+$validChars["\xA1\xB1"] = "\x00\x00\xFF\xE3";
+$fromUnicode["\xFF\xE3"] = "\xA1\xB1";
+
+/* Convert 0xA1EF to U+FFE5 (FULLWIDTH YEN SIGN), not U+00A5 (YEN SIGN) */
+$validChars["\xA1\xEF"] = "\x00\x00\xFF\xE5";
+$fromUnicode["\xFF\xE5"] = "\xA1\xEF";
+
+/* Convert U+00A5 (YEN SIGN) to 0x5C; that is one of the single bytes
+ * which many legacy Japanese text encodings used to represent something
+ * different from its normal meaning ASCII. In ASCII it's a backslash,
+ * but legacy Japanese software often used it for a yen sign. */
+$fromUnicode["\x00\xA5"] = "\x5C";
```

Comment:
Spent way to long reading about this and still not sure what the correct behavior is (especially as "correct per standard" and "correct in practice" don't necessarily agree here). If this is what we previously did, I guess we can keep doing it.

https://www.nslabs.jp/round-trip.rhtml seems like the best information on the topic.

Reply:
Thanks for the link. Just looking at it.

When I was working on all the SJIS variants, I also did some reading and thinking about the issue, compared the existing (mbstring) behavior for all questionable characters, compared how they are handled by other encoding conversion libraries, looked at what alternative representations are available in each SJIS variant, and decided on how to handle SJIS➡Unicode and Unicode➡SJIS for each variant.

For EUC-JP-2004, I just stuck with the existing behavior for expediency, although it was different from what was recommended in the conversion table which I found online (which, to be clear, is not really authoritative anyways).